include/mata/nfa/delta.hh

```diff
@@ -121,64 +121,105 @@ public:
     const_iterator find(const Symbol symbol) const { return super::find({ symbol, {} }); }
 
     /**
-     * Iterator over moves. It iterates over pairs (symbol, target state) for a current source state whose state post
-     *  we iterate.
+     * @brief Iterator over moves represented as @c Move instances.
+     *
+     * It iterates over pairs (symbol, target) for the given @c StatePost.
      */
-    struct moves_const_iterator {
-    private:
-        const std::vector<SymbolPost>& symbol_posts_{};
-        std::vector<SymbolPost>::const_iterator symbol_post_it_{};
-        StateSet::const_iterator target_states_it_{};
-        bool is_end_{ false };
-        Move move_{};
-
-    public:
-        using iterator_category = std::forward_iterator_tag;
-        using value_type = Move;
-        using difference_type = unsigned;
-        using pointer = Move*;
-        using reference = Move&;
-
-        explicit moves_const_iterator(const std::vector<SymbolPost>& symbol_posts, bool is_end = false);
-
-        moves_const_iterator(const std::vector<SymbolPost>& symbol_posts,
-                             std::vector<SymbolPost>::const_iterator symbol_posts_it,
-                             StateSet::const_iterator target_states_it, bool is_end = false);
-
-        moves_const_iterator(const moves_const_iterator& other) = default;
-
-        const Move& operator*() const { return move_; }
-
-        // Prefix increment
-        moves_const_iterator& operator++();
-        // Postfix increment
-        const moves_const_iterator operator++(int);
-
-        moves_const_iterator& operator=(const moves_const_iterator& x);
-
-        bool operator==(const moves_const_iterator& other) const;
-        bool operator!=(const moves_const_iterator& other) const { return !(*this == other); };
-    };
-
-    moves_const_iterator moves_cbegin() const { return moves_const_iterator(ToVector()); }
-    moves_const_iterator moves_cend() const { return moves_const_iterator(ToVector(), true); }
-    moves_const_iterator moves_begin() const { return moves_cbegin(); }
-    moves_const_iterator moves_end() const { return moves_cend(); }
-
     class Moves {
     public:
-        moves_const_iterator begin_;
-        moves_const_iterator end_;
-        moves_const_iterator begin() const { return begin_; }
-        moves_const_iterator end() const { return end_; }
-    };
+        /**
+         * @brief Whether to look up first symbol to iterate over from the beginning or from the end of @c StatePost
+         *  iterator.
+
+         * Iterates over symbol posts with symbols in interval [first_symbol_, last_symbol_].
+         */
+        enum class FirstSymbolLookupDiretion {
```

Comment:
```suggestion
        enum class FirstSymbolLookupDirection {
```
Maybe `LookupDirection` would suffice? Or `IterationDirection`?

Reply:
It is not an iteration direction. The moves are still iterated in the order of first to last, but the iteration starts at some first symbol which must be found in the vector of `SymbolPost`s first. The process of finding the iterator to this first symbol (and its `SymbolPost`) can proceed either backward (from the last element in the vector to the first one) or forward (from the first element to the last one). When you are looking for the first epsilon in the vector, you want to start from the end. Otherwise, you should start from the beginning of the vector.

Reply:
OK.

Reply:
We discussed and agreed to change it, this class will be gone, but I forgot what exactly should happen.

Reply:
We will have only one function, `StatePost::moves()` which will take 3 optional parameters. Without any, it will behave as a normal moves iterator iterating over all moves (including epsilons) and when the parameters are specified, it will iterate over some subset of these: `StatePost::moves(moves_kind, first_symbol_iterator, last_symbol_iterator)` where `moves_kind` is of type enum class with values `MovesKind::All`, `MovesKind::Epsilons` and `MovesKind::Symbols`. Iterators then only specify, which subset of moves is to be iterated over. The kind specifies whether to look for the first symbol from the left or from the right. This replaces `FirstSymbolLookupDirection` class.

Reply:
Sounds much better. Moreover, I though of few alternatives for `MovesKinds` (sounds kinda weird): `Pick::All`, `Select::Epsilons` or maybe `Iterate::Symbols`? (I wrote them with the second part as well, so one can see, how it sounds). Feel free to pick one, if you like any.

Reply:
I am experimenting with the names and `Iterate::Symbols` seems good to me. I will use `Iterate` and others can give their opinion in the PR when the changes are pushed.

Reply:
"Iterate" does not give much hint, no? 
But I don't know what it should be, IterateOver? SymbolKind / SymbolType....

Related to this, we should fix what "symbol" means. Currently we have the data type Symbol, that can be alphabet symbol or epsilon, but here David was using symbol as alphabet symbol only.
We could have symbol, alphabet symbol, epsilon, or perhaps better "Label" (for anything), "Symbol", and "Epsilon".

Reply:
I think it explains it well: `nfa.delta[4].moves(Iterate::Epsilons)` gives me all the information I need, in my opinion. I personally do not see a difference between `Iterate` and `IterateOver`. But I can change it to `IterateOver` if you think that is better. Either sounds good to me.

Regarding the `Symbol` naming. You said that you do not consider epsilon to be a symbol, and want the “normal” symbols to be called just symbols. Therefore, the name gets a bit ambiguous. I think we can think about possible names, discuss it tomorrow in our meeting and later refactor the whole Mata accordingly. For now, I will use “alphabet symbols” for “normal” symbols and “epsilons” for epsilon symbols.

Reply:
> "Iterate" does not give much hint, no? But I don't know what it should be, IterateOver? SymbolKind / SymbolType....

I think, here, you have to take into the account the context and the usage. Since, this will be probably called as `*.moves(Iterate::Symbols, ...)`, just simple verb with the noun is IMO fine.

Reply:
Hm, maybe one could use it, but is it sounds ilogical to me. If I read *.moves(bla) as a constructor of a set of moves 
(is this right? The set is not explicitly constructed, only iterable, but one still sees it as a set of moves.) 
then the bla should specify a property of the elements I want to include in the set. Iterate is not a porperty of an element, moves(Iterate:bla) would look like constructing the set of all moves with s special iteration that only considers its subset, weird.

Reply:
Whatever, I was thinking that it is iterator, and hence its parameters would specify properties of the iteration not its elements, but this discussion is getting nowhere, so go for the `MovesKind`.

Reply:
We agreed now to keep the original three methods to create `Moves`. Therefore, we do not need this parameter to pick which symbol posts we want to iterate over. This parameter will be changed back to `SeekFirstSymbolDirection::Forward` and will be used by the user only when they want to explicitly fill all parameters themselves: `moves(first_symbol, last_symbol, direction)`. Otherwise, in 99 % of cases, they will just use one of the three functions without needing to specify any parameter.

Reply:
You said to rename `Iterate` to `SeekFirst` or something similar earlier in our discussion in Discord. I can rename it to something else, but the meaning stays. It no longer represents a kind of moves to create. We have named functions for that. It only means where to start looking for the first symbol. What would you prefer?

Reply:
The `SeekFirst...` was replaced by a simple boolean flag parameter for constructor of `Moves`.

Reply:
Moved the logic getting begin and end symbol post iterators to `moves_...()` functions. Therefore, we cannot create a custom `Moves` class instance with a specified range of `[first_symbol, last_symbol]`. We can only create custom `Moves` class instance with iterators. If we want to specify the range with symbols, we need to use one of the helper functions for getting all labels, symbols, or epsilons.